core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -995,22 +1028,43 @@ protected boolean shouldExecuteReplication(Settings settings) {
         return IndexMetaData.isIndexUsingShadowReplicas(settings) == false;
     }
 
-    static class IndexShardReference implements Releasable {
+    interface IndexShardReference extends Releasable {
+        boolean isRelocated();
+
+        ShardRouting routingEntry();
+    }
+
+    static class IndexShardReferenceImpl implements IndexShardReference {
 
-        final private IndexShard counter;
+        private final IndexShard indexShard;
+        private final Releasable operationLock;
         private final AtomicBoolean closed = new AtomicBoolean();
 
-        IndexShardReference(IndexShard counter) {
-            counter.incrementOperationCounter();
-            this.counter = counter;
+        IndexShardReferenceImpl(IndexShard indexShard, boolean primaryAction) {
+            this.indexShard = indexShard;
+            if (primaryAction) {
+                operationLock = indexShard.acquirePrimaryOperationLock();
+            } else {
+                operationLock = indexShard.acquireReplicaOperationLock();
+            }
         }
 
         @Override
         public void close() {
             if (closed.compareAndSet(false, true)) {
```

Comment:
I think we don't need the AtomicBoolean here. We should protect from double closing on the level below